### ISSUE TITLE
Materialized view improvement

### ISSUE THREAD

**Issue by sungheun** [trinodb/trino#12466]
We have available and planned changes for materialized view improvement in AWS Athena and are going to port some changes below that are meaningful to Trino open source.

### Available changes
- [ ] Trigger an initial refresh during CREATE MATERIALIZED VIEW DDL execution. With the change, a user doesn't have to trigger an additional REFRESH command to exploit MV after MV is created.
- [ ] Add a new MV property to allow stale MV result access. 
- [ ] Allow source table type different from MV storage table

### Planned changes
- [ ] Hybrid plan: if there are data changes in the base table since last refresh, incremental result is generated only with the the changed data and it will be merged with existing storage table. The supported query shape/operator list will be limited. Additionally, it will be supported only for Iceberg because it will provide CDC feature. (https://github.com/apache/iceberg/pull/4539) More details will be followed later.
- [ ] Incremental refresh: More details will be followed later.

**Comment by raunaqmorarka** [comment]
> Trigger an initial refresh during CREATE MATERIALIZED VIEW DDL execution. With the change, a user doesn't have to trigger an additional REFRESH command to exploit MV after MV is created.

If we're changing CMVAS to populate the storage table (similar to CTAS) then we also need to provide a CMVAS WITH NO DATA so that MV can be defined without executing the query. It will also provide existing users a way to get back the previous behaviour.

> Add a new MV property to allow stale MV result access.

See related https://github.com/trinodb/trino/issues/11698
I think it's reasonable to add an MV property to iceberg like `stale-results-window` to specify the time interval from modification time of the source table upto which we will allow using stale storage table.

> The source table type that are different from MV storage table is allowed.

Could you clarify what this means ?

cc: @sopel39 @findepi @electrum @martint

**Comment by sopel39** [comment]
> The supported query shape/operator list will be limited. Additionally, it will be supported only for Iceberg because it will provide CDC feature. 

CDC should be part of SPI so that it's possible to run incremental refreshes from one connector (e.g. MySql) to another (e.g. Iceberg). This would enable us to integrate ETL ingestion natively with Trino federated model without depending on external tools to do the job of ingestion. For example, SPI should define mechanism for:
* table versioning
* versioning layout

My thinking about it that there should be set of methods in SPI:
```
Optional<ConnectorTableVersioningLayout> getTableVersioningLayout(ConnectorTableHandle handle);
class ConnectorTableVersioningLayout {
    /**
     * Specifies versioning granularity, e.g: per row (versioning on primary key), per partition, per table.
     **/
    Set<ColumnHandle> versioningGroupingSet;
}

/**
 * Returns current version of table (if present).
 */
OptionalLong getCurrentTableVersion(ConnectorTableHandle handle);

/**
 * Returns table handle that will provide data for new versions of versioning groups
 * (see ConnectorTableVersioningLayout#versioningGroupingSet).
 */
Optional<VersionApplicationResult> applyTableVersion(
  ConnectorTableHandle handle,
  // can be empty when it's first import
  OptionalLong fromVersionExclusive);
class VersioningApplicationResult {
  ConnectorTableHandle handle;
}

/**
 * Returns table handle that produces versioning grouping sets
 * (see ConnectorTableVersioningLayout#versioningGroupingSet)
 * that were removed from table since `fromVersionExclusive`.
 */
Optional<ConnectorTableHandle> getDeletedGroupingSets(ConnectorTableHandle handle, long fromVersionExclusive);
```
The model assumes that table versioning is global rather than independent per `ConnectorTableVersioningLayout#versioningGroupingSet`. I believe that global table versions are much more common (e.g. txid in OLTP, table version in Iceberg).

@sungheun it would be great if we worked together on incremental MVs

**Comment by sopel39** [comment]
> We have available and planned changes for materialized view improvement in AWS Athena and are going to port some changes below that are meaningful to Trino open source.

@sungheun do you mean MVs in Hive or Iceberg?

**Comment by sungheun** [comment]
> If we're changing CMVAS to populate the storage table (similar to CTAS) then we also need to provide a CMVAS WITH NO DATA so that MV can be defined without executing the query. It will also provide existing users a way to get back the previous behaviour.

This looks good to me. Anyway, default behavior of existing CMVAS will be changed to trigger an initial refresh.

> See related https://github.com/trinodb/trino/issues/11698
I think it's reasonable to add an MV property to iceberg like stale-results-window to specify the time interval from modification time of the source table upto which we will allow using stale storage table.

I am not sure if every connector can provide information when the source table was updated last time. In this context, I want to introduce more generic property that can be applied to every connector.

>> The source table type that are different from MV storage table is allowed.

> Could you clarify what this means ?

Current MV implementation of Iceberg connector assumes that the source table is also Iceberg table to retrieve transaction snapshot IDs. Want to remove the limitation so that other table type such as Hive are allowed. This implementation will be meaningful only for an MV that allows stale result.

Original issue description will be also updated to apply the feedback.
@raunaqmorarka

**Comment by sungheun** [comment]
@sopel39 
I am definitely happy to work together on this topic.
Let me answer your comments/questions below.

- I fully agree with you that CDC should be part of SPI interface so that any connector can implement the interface if it has a capability.
 - The basic idea of the interface that you described looks good to me even though I have some questions on the details. Can you make a PR (or another issue) so that we can further discuss the details?
- On top of the interface, we need to implement incremental result generation in the optimizer side. I have a prototyping code that is not verified yet. Will share it later once ready, but proper SPI interface definition is a task that needs to be done prior to this.
- In the first step, we are thinking of supporting Hive & Iceberg table. In the future, it can be further extended.

**Comment by sopel39** [comment]
> The basic idea of the interface that you described looks good to me even though I have some questions on the details. Can you make a PR (or another issue) so that we can further discuss the details?

@sungheun https://github.com/trinodb/trino/pull/12499

**Comment by zhengxingmao** [comment]
I verified the materialized view under the small and medium data amounts in the 375 and 386 versions, and found that the data will be automatically refreshed, is the automatic refresh mechanism fully supported?

**Comment by lozbrown** [comment]
Will this result in the creation of hive materialized views being possible? 

if i'm not mistaken hive materialized views are not currently supported...

**Comment by sopel39** [comment]
> Will this result in the creation of hive materialized views being possible?

It will be more about Iceberg MVs TBH. However, SPI will existing in any connector for use
